File: WinPaletter/GUI/Elements/Checkers/RadioImage.vb
Imports System.ComponentModel
Imports System.Drawing.Drawing2D
Imports System.Drawing.Text

Namespace UI.WP

    <Description("RadioButton, but with image for WinPaletter UI")> <DefaultEvent("CheckedChanged")> Public Class RadioImage : Inherits Control

        Sub New()
            SetStyle(DirectCast(139286, ControlStyles), True)
            SetStyle(ControlStyles.Selectable, True)
            DoubleBuffered = True
            Font = New Font("Segoe UI", 9)
            ForeColor = Color.White
            Text = ""
        End Sub

#Region "Variables"

        Private AnimateOnClick As Boolean = False

        Public State As MouseState = MouseState.None

        Enum MouseState
            None
            Over
            Down
        End Enum

#End Region

#Region "Properties"

        Private _Checked As Boolean
        Public Property Checked() As Boolean
            Get
                Return _Checked
            End Get
            Set(value As Boolean)
                Try
                    _Checked = value

                    If _Checked Then
                        UncheckOthersOnChecked()
                    End If

                    RaiseEvent CheckedChanged(Me)

                    Invalidate()

                Catch
                End Try
            End Set
        End Property

        Public Property Image As Image
        Public Property ShowText As Boolean = False

        <Browsable(True)>
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)>
        <EditorBrowsable(EditorBrowsableState.Always)>
        <Editor(GetType(System.ComponentModel.Design.MultilineStringEditor), GetType(System.Drawing.Design.UITypeEditor))>
        <Bindable(True)>
        Public Overrides Property Text As String = ""

        Public Property TextAlign As ContentAlignment = ContentAlignment.MiddleCenter
#End Region

#Region "Events"

        Event CheckedChanged(sender As Object)

        Protected Overrides Sub OnDragOver(e As DragEventArgs)
            If Not Checked AndAlso TypeOf e.Data.GetData(GetType(UI.Controllers.ColorItem).FullName) Is UI.Controllers.ColorItem Then
                e.Effect = DragDropEffects.None
                Checked = True
            Else
                Exit Sub
            End If
            MyBase.OnDragOver(e)
        End Sub

        Protected Overrides Sub OnMouseDown(e As MouseEventArgs)
            AnimateOnClick = True
            Checked = True
            State = MouseState.Down
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
            MyBase.OnMouseDown(e)
        End Sub

        Protected Overrides Sub OnMouseUp(e As MouseEventArgs)
            State = MouseState.Over
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
        End Sub

        Private Sub RadioImage_MouseEnter(sender As Object, e As EventArgs) Handles Me.MouseEnter
            State = MouseState.Over
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
        End Sub

        Private Sub RadioImage_MouseLeave(sender As Object, e As EventArgs) Handles Me.MouseLeave
            State = MouseState.None
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
        End Sub

        Private Sub RadioImage_HandleCreated(sender As Object, e As EventArgs) Handles Me.HandleCreated
            Try
                If Not DesignMode Then
                    AddHandler FindForm.Shown, AddressOf Showed
                    AddHandler Parent.BackColorChanged, AddressOf RefreshColorPalette
                    AddHandler Parent.VisibleChanged, AddressOf RefreshColorPalette
                    AddHandler Parent.EnabledChanged, AddressOf RefreshColorPalette
                    AddHandler VisibleChanged, AddressOf RefreshColorPalette
                    AddHandler EnabledChanged, AddressOf RefreshColorPalette
                End If
            Catch
            End Try

            Try
                alpha = 0
            Catch
            End Try
        End Sub

        Private Sub RadioImage_HandleDestroyed(sender As Object, e As EventArgs) Handles Me.HandleDestroyed
            Try
                If Not DesignMode Then
                    RemoveHandler FindForm.Shown, AddressOf Showed
                    RemoveHandler Parent.BackColorChanged, AddressOf RefreshColorPalette
                    RemoveHandler Parent.VisibleChanged, AddressOf RefreshColorPalette
                    RemoveHandler Parent.EnabledChanged, AddressOf RefreshColorPalette
                    RemoveHandler VisibleChanged, AddressOf RefreshColorPalette
                    RemoveHandler EnabledChanged, AddressOf RefreshColorPalette
                End If
            Catch
            End Try
        End Sub

        Sub Showed(sender As Object, e As EventArgs)
            Invalidate()
        End Sub

        Public Sub RefreshColorPalette(sender As Object, e As EventArgs)
            Invalidate()
        End Sub

#End Region

#Region "Subs/Functions"
        Private Sub UncheckOthersOnChecked()
            If Parent Is Nothing Then Return

            For Each C As Control In Parent.Controls
                If Not (C Is Me) AndAlso (TypeOf C Is RadioImage) Then
                    DirectCast(C, RadioImage).Checked = False
                End If
            Next
        End Sub

#End Region

#Region "Animator"

        Dim alpha As Integer
        ReadOnly Factor As Integer = 25
        Dim WithEvents Timer As New Timer With {.Enabled = False, .Interval = 1}

        Private Sub Timer_Tick(sender As Object, e As EventArgs) Handles Timer.Tick
            If Not DesignMode Then

                If State = MouseState.Over Then
                    If alpha + Factor <= 255 Then
                        alpha += Factor
                    ElseIf alpha + Factor > 255 Then
                        alpha = 255
                        Timer.Enabled = False
                        Timer.Stop()
                        AnimateOnClick = False
                    End If

                    Threading.Thread.Sleep(1)
                    Invalidate()
                End If

                If Not State = MouseState.Over Then
                    If alpha - Factor >= 0 Then
                        alpha -= Factor
                    ElseIf alpha - Factor < 0 Then
                        alpha = 0
                        Timer.Enabled = False
                        Timer.Stop()
                        AnimateOnClick = False
                    End If

                    Threading.Thread.Sleep(1)
                    Invalidate()
                End If
            End If
        End Sub

#End Region

        Protected Overrides Sub OnPaint(e As PaintEventArgs)
            Try
                Dim G As Graphics = e.Graphics
                If Parent Is Nothing Then Exit Sub
                G.Clear(GetParentColor)

                G = e.Graphics
                G.SmoothingMode = SmoothingMode.AntiAlias
                G.TextRenderingHint = If(DesignMode, TextRenderingHint.ClearTypeGridFit, TextRenderingHint.SystemDefault)
                DoubleBuffered = True

                Dim MainRect As New Rectangle(0, 0, Width - 1, Height - 1)
                Dim MainRectInner As New Rectangle(1, 1, Width - 3, Height - 3)
                Dim TextRect As New Rectangle(5, 5, Width - 10, Height - 10)

                Dim CenterRect As New Rectangle

                If Image IsNot Nothing Then CenterRect = New Rectangle(MainRect.X + (MainRect.Width - Image.Width) / 2,
                                            MainRect.Y + (MainRect.Height - Image.Height) / 2,
                                            Image.Width, Image.Height)

                Dim bkC As Color = If(_Checked, My.Style.Colors.Back_Checked, My.Style.Colors.Back)
                Dim bkCC As Color = Color.FromArgb(alpha, My.Style.Colors.Back_Checked)

                Using br As New SolidBrush(bkC) : G.FillRoundedRect(br, MainRectInner) : End Using
                Using br As New SolidBrush(bkCC) : G.FillRoundedRect(br, MainRect) : End Using

                Dim lC As Color = Color.FromArgb(255 - alpha, If(_Checked, My.Style.Colors.Border_Checked, My.Style.Colors.Border))
                Dim lCC As Color = Color.FromArgb(alpha, My.Style.Colors.Border_Checked_Hover)

                Using P As New Pen(lC) : G.DrawRoundedRect_LikeW11(P, MainRectInner) : End Using
                Using P As New Pen(lCC) : G.DrawRoundedRect_LikeW11(P, MainRect) : End Using

                If Image IsNot Nothing Then G.DrawImage(Image, CenterRect)

                If ShowText Then
                    Using br As New SolidBrush(ForeColor) : G.DrawString(Text, Font, br, TextRect, TextAlign.ToStringFormat) : End Using
                End If
            Catch

            End Try
        End Sub

    End Class

End Namespace

File: WinPaletter/GUI/Elements/Checkers/RadioImage.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Text;
using System.Windows.Forms;

namespace WinPaletter.UI.WP
{
    [Description("RadioButton, but with image for WinPaletter UI")]
    [DefaultEvent("CheckedChanged")]
    public class RadioImage : Control
    {
        public RadioImage()
        {
            SetStyle(ControlStyles.UserPaint | ControlStyles.SupportsTransparentBackColor | (ControlStyles)139286 | ControlStyles.Selectable, true);
            DoubleBuffered = true;
            BackColor = Color.Transparent;

            Font = new("Segoe UI", 9f);
            ForeColor = Color.White;
            Text = string.Empty;

            _alpha = 0; _alpha2 = _checked ? 255 : 0;
        }

        #region Variables
        private bool CanAnimate => !DesignMode && Program.Style.Animations && this != null && Visible && Parent != null && Parent.Visible && FindForm() != null && FindForm().Visible;

        public MouseState State = MouseState.None;

        public enum MouseState
        {
            None,
            Over,
            Down
        }

        #endregion

        #region Properties
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(false)]
        public new Color BackColor { get => Color.Transparent; set {; } }

        private bool _checked;
        public bool Checked
        {
            get => _checked;
            set
            {
                if (value != _checked)
                {
                    _checked = value;
                    if (_checked) { UncheckOthersOnChecked(); }

                    if (CanAnimate)
                    {
                        FluentTransitions.Transition.With(this, nameof(alpha2), _checked ? 255 : 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration));
                    }
                    else { alpha2 = Checked ? 255 : 0; }

                    CheckedChanged?.Invoke(this, new EventArgs());
                }
            }
        }


        private Image _image; 
        private Image _imageDisabled;
        public Image Image
        {
            get => Enabled ? _image : _imageDisabled;
            set
            {
                if (value != _image)
                {
                    _image = value;
                    _imageDisabled = value?.Grayscale() ?? null;
                    Invalidate();
                }
            }
        }


        [Browsable(true)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        [EditorBrowsable(EditorBrowsableState.Always)]
        [Editor(typeof(System.ComponentModel.Design.MultilineStringEditor), typeof(System.Drawing.Design.UITypeEditor))]
        [Bindable(true)]
        public override string Text { get; set; } = string.Empty;

        public ContentAlignment TextAlign { get; set; } = ContentAlignment.MiddleCenter;
        public ContentAlignment ImageAlign { get; set; } = ContentAlignment.MiddleCenter;
        public TextImageRelation TextImageRelation { get; set; } = TextImageRelation.ImageAboveText;

        #endregion

        #region Events/Overrides

        public event EventHandler CheckedChanged;

        protected override void OnDragOver(DragEventArgs e)
        {
            if (!Checked && e.Data.GetData(typeof(Controllers.ColorItem).FullName) is Controllers.ColorItem)
            {
                e.Effect = DragDropEffects.None;
                Checked = true;
            }
            else
            {
                return;
            }

            base.OnDragOver(e);
        }

        protected override void OnMouseDown(MouseEventArgs e)
        {
            Checked = true;
            State = MouseState.Down;

            if (CanAnimate) { FluentTransitions.Transition.With(this, nameof(alpha), 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = 0; }

            base.OnMouseDown(e);
        }

        protected override void OnMouseUp(MouseEventArgs e)
        {
            State = MouseState.Over;

            if (CanAnimate) { FluentTransitions.Transition.With(this, nameof(alpha), ContainsFocus ? 255 : 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = ContainsFocus ? 255 : 0; }

            base.OnMouseUp(e);
        }

        protected override void OnMouseEnter(EventArgs e)
        {
            State = MouseState.Over;

            if (CanAnimate) { FluentTransitions.Transition.With(this, nameof(alpha), 255).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = 255; }

            base.OnMouseEnter(e);
        }

        protected override void OnMouseLeave(EventArgs e)
        {
            State = MouseState.None;

            if (CanAnimate) { FluentTransitions.Transition.With(this, nameof(alpha), 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = 0; }

            base.OnMouseLeave(e);
        }

        protected override void Dispose(bool disposing)
        {
            base.Dispose(disposing);

            _image?.Dispose();
        }

        int parentLevel = 0;
        protected override void OnParentChanged(EventArgs e)
        {
            base.OnParentChanged(e);

            parentLevel = this.Level();
        }

        #endregion

        #region Methods
        private void UncheckOthersOnChecked()
        {
            if (Parent is null)
                return;

            foreach (Control C in Parent.Controls)
            {
                if (C != this && C is RadioImage)
                {
                    ((RadioImage)C).Checked = false;
                }
            }
        }

        #endregion

        #region Animator
        private int _alpha = 0;

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(false)]
        public int alpha
        {
            get => _alpha;
            set { _alpha = value; Invalidate(); }
        }

        private int _alpha2 = 0;

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(false)]
        public int alpha2
        {
            get => _alpha2;
            set { _alpha2 = value; Invalidate(); }
        }
        #endregion

        protected override void OnPaintBackground(PaintEventArgs pevent)
        {
            //Leave it empty to make control background transparent
            base.OnPaintBackground(pevent);
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            Graphics G = e.Graphics;
            G.SmoothingMode = SmoothingMode.AntiAlias;
            G.TextRenderingHint = DesignMode ? TextRenderingHint.ClearTypeGridFit : Program.Style.TextRenderingHint;

            Rectangle MainRect = new(0, 0, Width - 1, Height - 1);
            Rectangle MainRectInner = new(1, 1, Width - 3, Height - 3);
            Rectangle PaddingRect = Rectangle.FromLTRB(Padding.Left, Padding.Top, Padding.Right, Padding.Bottom);
            Rectangle TextAndImageRect = new(3 + PaddingRect.X, 3 + PaddingRect.Y, Width - 7 - PaddingRect.Width * 2, Height - 7 - PaddingRect.Height * 2);

            //Makes background drawn properly, and transparent
            InvokePaintBackground(this, e);

            Config.Scheme scheme = Enabled ? Program.Style.Schemes.Main : Program.Style.Schemes.Disabled;

            Color back = Color.FromArgb(alpha, scheme.Colors.Back_Checked_Hover);
            Color line = Color.FromArgb(255 - alpha, _checked ? scheme.Colors.Line_Checked : State != MouseState.Over ? scheme.Colors.Line(parentLevel) : scheme.Colors.Line_Checked_Hover);
            Color line_hover = Color.FromArgb(alpha, scheme.Colors.Line_Checked_Hover);

            using (LinearGradientBrush br = new(MainRect, Color.FromArgb(_alpha2, scheme.Colors.Back_Checked), Color.FromArgb(_alpha2, scheme.Colors.Back_Checked_Hover), LinearGradientMode.ForwardDiagonal))
            {
                G.FillRoundedRect(br, MainRect);
            }

            using (SolidBrush br = new(Color.FromArgb(255 - _alpha2, scheme.Colors.Back(parentLevel)))) { G.FillRoundedRect(br, MainRectInner); }

            using (SolidBrush br = new(back)) { G.FillRoundedRect(br, MainRect); }

            using (Pen P = new(line)) { G.DrawRoundedRectBeveledReverse(P, MainRectInner); }

            using (Pen P = new(line_hover)) { G.DrawRoundedRectBeveledReverse(P, MainRect); }

            this.GetTextAndImageRectangles(TextAndImageRect, out RectangleF imageRectF, out RectangleF textRectF);

            // Draw image
            if (Image != null) G.DrawImage(Image, Rectangle.Round(imageRectF));

            // Draw text
            if (!string.IsNullOrEmpty(Text))
            {
                if (TextAndImageRect.Height % 2 == 0) textRectF.Y += 0.5f;

                if (Enabled)
                {
                    TextRenderer.DrawText(G, Text, Font, Rectangle.Round(textRectF), ForeColor, TextFormatFlags.EndEllipsis | TextFormatFlags.VerticalCenter);
                }
                else
                {
                    Rectangle disabledTextRect = Rectangle.Round(textRectF);
                    disabledTextRect.X--; disabledTextRect.Y--;
                    ControlPaint.DrawStringDisabled(G, Text, Font, ForeColor, disabledTextRect, TextFormatFlags.EndEllipsis | TextFormatFlags.VerticalCenter);
                }
            }

            base.OnPaint(e);
        }
    }
}